src/pymordemos/fin_optimization.py

```diff
@@ -0,0 +1,247 @@
+            der_expr[key][i] = '{}*{}**2*({}[{}]-'.format(weights[key][i], 1., key, i) + '{}'.format(mu_d[key][i]) + ')'
+            second_der_expr[key][i][key][i] = '{}*{}**2'.format(weights[key][i], 1.)
+            parameter_functionals.append(
+                ExpressionParameterFunctional(
+                    '{}*{}**2*0.5*({}[{}]'.format(weights[key][i], 1., key, i)
+                    + '-{}'.format(mu_d[key][i])+')**2', second_derivative_expressions=second_der_expr,
+                    derivative_expressions=der_expr, parameters=helper_problem.parameters))
+    coeffs = []
+    for f in parameter_functionals:
+        coeffs.append(1.)
+    const_coeff = LincombParameterFunctional(parameter_functionals, coeffs)
+    lin_coeff = -1. * T_root_d.to_numpy()[0][0]
+    const_op = ConstantOperator(NumpyVectorSpace(1).ones(), boundary_functional.range)
+    bilin_op = NumpyMatrixOperator(
+        diags(boundary_functional.H.matrix[0]), source_id=boundary_functional.range.id,
+        range_id=boundary_functional.range.id
+    )
+    output_functional = OutputFunctional(
+        {'constant': [const_op], 'linear': [boundary_functional.H], 'bilinear': [bilin_op]},
+        {'constant': [const_coeff], 'linear': [lin_coeff], 'bilinear': [.5]}
+    )
+    l2_boundary_product = RobinBoundaryOperator(
+        grid, data['boundary_info'], name='l2_boundary_product',
+        robin_data=(ConstantFunction(1, 2), ConstantFunction(1, 2))
+    )
+    helper_fom = helper_fom.with_(
+        products=dict(
+            opt=helper_fom.energy_product, l2_boundary=l2_boundary_product,
+            **helper_fom.products), output_functional=output_functional
+    )
+    pde_opt_fom = StationaryModel(
+        helper_fom.operator, helper_fom.rhs, output_functional=output_functional,
+        products=helper_fom.products, error_estimator=helper_fom.error_estimator,
+        visualizer=helper_fom.visualizer
+    )
+    range_1 = helper_problem.parameter_space.ranges['biot']
+    range_2 = helper_problem.parameter_space.ranges['k0']
+    return pde_opt_fom, parameter_space, mu_bar, range_1, range_2
+
+
+def record_results(function, data, mu=None):
+    QoI = function(mu)
+    data['num_evals'] += 1
+    data['evaluation_points'].append(mu)
+    data['evaluations'].append(QoI)
+    return QoI
+
+
+def report(result, parse, data, reference_mu=None):
+    if (result.status != 0):
+        print('\n failed!')
+    else:
+        print('\n succeeded!')
+        print(f'  mu_min:    {parse(result.x)}')
+        print(f'  J(mu_min): {result.fun}')
+        print(f'  absolute error w.r.t. reference solution: {np.linalg.norm(result.x-reference_mu):.2e}')
+        print(f'  num iterations:     {result.nit}')
+        print(f'  num function calls: {data["num_evals"]}')
+        print(f'  time:               {data["time"]:.5f} seconds')
+        if 'offline_time' in data:
+            print(f'  offline time:       {data["offline_time"]:.5f} seconds')
+    print('')
+
```

Comment:
The two functions above can be imported from the `linear_optimization` demo, I'd say.